packages/server/lib/browsers/index.ts

```diff
@@ -129,89 +121,81 @@ exports = {
+    const browserLauncher = await getBrowserLauncher(browser, options.browsers)
 
     await browserLauncher.connectToExisting(browser, options, automation)
 
     return this.getBrowserInstance()
   },
 
   async connectToNewSpec (browser: Browser, options: BrowserNewTabOpts, automation: Automation) {
-    const browserLauncher = getBrowserLauncher(browser, options.browsers)
+    const browserLauncher = await getBrowserLauncher(browser, options.browsers)
 
     // Instance will be null when we're dealing with electron. In that case we don't need a browserCriClient
     await browserLauncher.connectToNewSpec(browser, options, automation)
 
     return this.getBrowserInstance()
   },
 
-  open (browser: Browser, options: BrowserLaunchOpts, automation: Automation, ctx) {
-    return kill(true)
-    .then(() => {
-      _.defaults(options, {
-        onBrowserOpen () {},
-        onBrowserClose () {},
-      })
-
-      ctx.browser.setBrowserStatus('opening')
-
-      const browserLauncher = getBrowserLauncher(browser, options.browsers)
-
-      if (!options.url) throw new Error('Missing url in browsers.open')
-
-      debug('opening browser %o', browser)
-
-      return browserLauncher.open(browser, options.url, options, automation)
-      .then((i) => {
-        debug('browser opened')
-        // TODO: bind to process.exit here
-        // or move this functionality into cypress-core-launder
-
-        i.browser = browser
-
-        instance = i
-
-        // TODO: normalizing opening and closing / exiting
-        // so that there is a default for each browser but
-        // enable the browser to configure the interface
-        instance.once('exit', () => {
-          ctx.browser.setBrowserStatus('closed')
-          // TODO: make this a required property
-          if (!options.onBrowserClose) throw new Error('onBrowserClose did not exist in interactive mode')
-
-          options.onBrowserClose()
-          instance = null
-        })
-
-        // TODO: instead of waiting an arbitrary
-        // amount of time here we could instead
-        // wait for the socket.io connect event
-        // which would mean that our browser is
-        // completely rendered and open. that would
-        // mean moving this code out of here and
-        // into the project itself
-        // (just like headless code)
-        // ----------------------------
-        // give a little padding around
-        // the browser opening
-        return Promise.delay(1000)
-        .then(() => {
-          if (instance === null) {
-            return null
-          }
-
-          // TODO: make this a required property
-          if (!options.onBrowserOpen) throw new Error('onBrowserOpen did not exist in interactive mode')
-
-          options.onBrowserOpen()
-          ctx.browser.setBrowserStatus('open')
-
-          return instance
-        })
-      })
+  async open (browser: Browser, options: BrowserLaunchOpts, automation: Automation, ctx) {
```

Comment:
```suggestion
  async open (browser: Browser, options: BrowserLaunchOpts, automation: Automation, ctx): BrowserInstance {
```

Reply:
Currently, these 3 methods can actually resolve with `null | BrowserInstance`, since the browser can theoretically exit and null out `instance`. Added explicit types here: 5e743998a8